File: sampling/cxx/test_sampling.cxx
#include "simple_random_sample.cxx"
#include <cassert>
#include <cstdlib>
#include <vector>

typedef std::vector<int> (*SamplingFunc)(const std::vector<int>&, int);

void test_srs(SamplingFunc func)
{
	const int SIZE = 10;
	std::vector<int> array(SIZE, 0);
	for(int i = 0; i < array.size(); i++)
		array[i] = i;
	std::vector<int> count(SIZE, 0);
	const int TIMES = 1000;
	for(int i = 0; i < TIMES; i++)
	{
		std::vector<int> sample = func(array, SIZE / 2);
		for(const auto& num: sample)
			count[num] += 1;
	}
	for(int i = 0; i < SIZE; i++)
	{
		assert(0.4 * TIMES < count[i]);
		assert(count[i] < 0.6 * TIMES);
	}
}

int main()
{
	test_srs(selection_rejection);
	test_srs(reservoir_sampling);
	return 0 ;
}

File: sampling/py3/TestSampling.py
import unittest
from collections import Counter
from simple_random_sample import *
from weightd_random_sample import weighted_random_sample

class TestSampling(unittest.TestCase):
    def __test_srs__(self, algo, isIter):
        size = 10
        array = list(range(size))
        cnt = Counter()
        times = 1000
        for i in range(times):
            if isIter:
                sample = algo(iter(array), size // 2)
            else:
                sample = algo(array, size // 2)
            self.assertEqual(len(sample), size // 2)
            for num in sample:
                cnt[num] += 1
        for num in array:
            self.assertGreater(cnt[num], 0.4 * times)
            self.assertLess(cnt[num], 0.6 * times)

    def testWeightedSampling(self):
        size = 1000
        for i in range(100):
            result = sum(weighted_random_sample([1, 2]) for i in range(size))
            self.assertLessEqual(result, size * 0.72)
            self.assertGreaterEqual(result, size * 0.62)

    def testDrawByDraw(self):
        self.__test_srs__(draw_by_draw, False)

    def testSelectionRejection(self):
        self.__test_srs__(selection_rejection, False)

    def testReservoirSampling(self):
        self.__test_srs__(reservoir_sampling, True)

if __name__ == "__main__":
    unittest.main()
